### PR TITLE
Raylet request resource report endpoint

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This adds a raylet endpoint for GCS to pull resource usage information. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -451,11 +451,10 @@ void NodeManager::Heartbeat() {
   });
 }
 
-void NodeManager::ReportResourceUsage() {
-  auto resources_data = std::make_shared<rpc::ResourcesData>();
+void NodeManager::FillResourceReport(rpc::ResourcesData *resources_data) {
   resources_data->set_node_id(self_node_id_.Binary());
   resources_data->set_node_manager_address(initial_config_.node_manager_address);
-  // Update local chche from gcs remote cache, this is needed when gcs restart.
+  // Update local chache from gcs remote cache, this is needed when gcs restart.
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:457]:
```suggestion
  // Update local cache from gcs remote cache, this is needed when gcs restart.
```

File: src/ray/raylet/node_manager.h [hunk 2]
```diff
@@ -696,7 +705,7 @@ class NodeManager : public rpc::NodeManagerServiceHandler,
   ///
   /// \param Output parameter. `resource_load` and `resource_load_by_shape` are the only
   /// fields used.
-  void FillResourceUsage(std::shared_ptr<rpc::ResourcesData> data) override;
+  void FillResourceUsage(rpc::ResourcesData *data) override;
```
Comment (ericl) [inline src/ray/raylet/node_manager.h:708]: Why change to a raw ptr?

Comment (wuisawesome) [reply src/ray/raylet/node_manager.h:708]: Because grpc gives raw pointers (like `request_resource_report_reply->mutable_resources()`)

**Review by ericl — APPROVED** [review]
Please fix the spelling mistake

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -451,11 +451,10 @@ void NodeManager::Heartbeat() {
   });
 }
 
-void NodeManager::ReportResourceUsage() {
-  auto resources_data = std::make_shared<rpc::ResourcesData>();
+void NodeManager::FillResourceReport(rpc::ResourcesData *resources_data) {
```
Comment (iycheng) [inline src/ray/raylet/node_manager.cc:454]: Maybe pass by reference? I feel that should be a better practice since it can't be null in this case.

**Review by fishbone — COMMENTED** [review]
LGTM. The only place I have concern is about passing by raw ptr. Usually, if the function always treats it as non-null ptr, passing by reference should be better.

File: src/ray/raylet/node_manager.h [hunk 4]
```diff
@@ -543,6 +547,11 @@ class NodeManager : public rpc::NodeManagerServiceHandler,
   /// \return Status indicating whether setup was successful.
   ray::Status SetupPlasmaSubscription();
 
+  /// Handle a `RequestResourceReport`A
```
Comment (clarkzinzow) [inline src/ray/raylet/node_manager.h:550]:
```suggestion
  /// Handle a `RequestResourceReport` request.
```

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -1458,6 +1462,15 @@ void NodeManager::ProcessSubmitTaskMessage(const uint8_t *message_data) {
   SubmitTask(Task(task_message));
 }
 
+void NodeManager::HandleRequestResourceReport(
+    const rpc::RequestResourceReportRequest &request,
+    rpc::RequestResourceReportReply *reply, rpc::SendReplyCallback send_reply_callback) {
+  auto resources_data = reply->mutable_resources();
+  FillResourceReport(resources_data);
```
Comment (clarkzinzow) [inline src/ray/raylet/node_manager.cc:1469]: Nit: Any reason to not pass the resources pointer straight in as an rvalue?

```suggestion
  FillResourceReport(reply->mutable_resources());
```

**Review by clarkzinzow — APPROVED** [review]

**Comment by clarkzinzow** [comment]
@wuisawesome This PR is superseded by #14336, right? If that's the case, can you close this PR?

**Comment by wuisawesome** [comment]
I prefer to merge the raylet side first to isolate the pull logic, which still needs additional testing.

**Review by fishbone — APPROVED** [review]

**Comment by ericl** [comment]
Any progress on this?

**Comment by wuisawesome** [comment]
git really screwed up merging this with the old scheduler removal pr. i'll get this merged by eod

**Pull request merged**